pkg/controller/constrainttemplate/constrainttemplate_controller.go

```diff
@@ -371,3 +396,80 @@ func removeString(s string, items []string) []string {
 	}
 	return rval
 }
+
+// RemoveAllFinalizers removes all finalizers from constraints and
+// constraint templates
+func RemoveAllFinalizers(c client.Client, finished chan struct{}) {
+	defer close(finished)
+	toList := func(m map[types.NamespacedName]string) []string {
+		var out []string
+		for k := range m {
+			out = append(out, k.Name)
+		}
+		return out
+	}
+
+	templates := &v1beta1.ConstraintTemplateList{}
+	names := make(map[types.NamespacedName]string)
+	if err := c.List(context.Background(), nil, templates); err != nil {
+		log.Error(err, "could not clean all contraint/template finalizers")
+		return
+	}
+	for _, templ := range templates.Items {
+		names[types.NamespacedName{Name: templ.GetName()}] = templ.Spec.CRD.Spec.Names.Kind
+	}
+	log.Info("found constraint templates to scrub", "templates", toList(names))
+	cleanLoop := func() (bool, error) {
+		log.Info("removing finalizers from constraint templates", "templates", toList(names))
+		for nn, kind := range names {
+			listKind := kind + "List"
+			// TODO these should be constants somewhere
+			gvk := schema.GroupVersionKind{Group: "constraints.gatekeeper.sh", Version: "v1beta1", Kind: listKind}
+			objs := &unstructured.UnstructuredList{}
+			objs.SetGroupVersionKind(gvk)
+			if err := c.List(context.Background(), nil, objs); err != nil {
+				log.Error(err, "while listing constraints for cleanup", "kind", listKind)
+				continue
+			}
+			success := true
```

Comment:
Would it make sense to set `success := false` first then set to `true` if finalizer has been removed?

Reply:
It needs to be `success := true`, so that a single failure turns the attempt to clean the kind unsuccessful. I don't think there is a race condition here b/c the code removing Constraint and CT finalizers is single-threaded.

Reply:
Alternately, the sync cleaner uses `failure := false`. This is the same as `success := true` but may read easier.